File: Erlang/customer.erl
%% @author harshkour1
%% @doc @todo Add description to customer.


-module(customer).

%% ====================================================================
%% API functions
%% ====================================================================
-export([startcust/3]).
-import(lists,[nth/2]).
-import(lists,[delete/2]).
-import(timer,[sleep/1]).
-export([for/6]).
-export([custreceive/1]).
-import(money,[requestreceive/0]).
-import(bank,[bankreq/2]).


%% ====================================================================
%% Internal functions
%% ====================================================================


startcust(Name, Amount, Bkeys) ->
	Length = length(Bkeys),
	for(Length,Name,Amount,Amount,Amount,Bkeys),
	custreceive(Name).

custreceive(Name) ->
				receive
					{approve, Name, Bname, Famt,Tamt, Amount,Keysb} -> 
							cust_update(1,Name,Bname,Famt,Tamt, Amount,Keysb),
							custreceive(Name);
					{reject, Name, Bname,Famt, Tamt, Amount,Keysb} -> 
							cust_update(2,Name,Bname,Famt,Tamt, Amount,Keysb),
							custreceive(Name)
				end.
	
cust_update(Flag,Name,Bname,Famt,Tamt,Amount,Keysb) ->
	if
		Flag == 1 ->
			Amount1 = Tamt - Amount,
			Keysbb= Keysb,
			Length = length(Keysbb);	
		true ->
			Amount1 = Tamt,
			Keysbb = delete(Bname,Keysb),
			Length = length(Keysbb),
			io:fwrite("")
	end,
	if Length == 0 andalso Flag == 2->
		 Amtleft = Famt - Tamt,
		whereis(money) ! {borrowleft,Name,Amtleft};   
	true ->
		io:fwrite("")
	end,
	if Amount1 == 0 ->
		 whereis(money) ! {reached,Name,Famt};
	true ->
		io:fwrite("")
	end,
	if
		Amount1 > 0 andalso Length /= 0 ->
			F =  rand:uniform(Length),
			for(F,Name,Famt,Amount1,Amount1,Keysbb);
		true ->
			io:fwrite("")
	end.


for(Max,Name,Famount,Tamount, Amount,Bkeys) ->
	Amount1 = Amount,
	random:seed(now()),
	D = rand:uniform(100),
	if D > 10 ->
		timer:sleep(rand:uniform(D));
	true ->
		timer:sleep(10)
	end,
	if 
		Amount1 > 50 ->
			A =  rand:uniform(50);
		true ->
			A =  rand:uniform(Amount1)
	end,
	
	Bankcalled = nth(rand:uniform(Max),Bkeys),
	Bankpid = whereis(Bankcalled),
	whereis(money) ! {request,Name,A,Bankcalled},
	Bankpid ! {Name, Famount,Tamount, A, Bkeys}.
		
	

File: Java/customer.java
//@author:harshkour
import java.util.*;
public class customer implements Runnable{

	private String custname;
	private int custamount;
	Thread t2;
	money m;
	List<String> banknames;
	static List<bank> banks1;
	List<bank> banks;
	static int count = 0;
	customer(String custname,int custamount, List<String> banknames,money m){
		this.custname = custname;
		this.custamount = custamount;
		this.m = m;
		this.banknames = banknames;
		t2 = new Thread(this, custname);
	}
	
	customer(){
		
	}
	public void bankobjects(List<bank> banklist) {
		this.banks = new ArrayList<>(banklist);
			banks1 = new ArrayList<>(banklist);
		//System.out.println(this.banks);
	}
	
	public void custobjects(List<customer> custlist) {
		count = custlist.size();
		//System.out.println("count is "+ count);
		//for(int i= 0 ;i < count; i++) {
		//	customer c = custlist.get(i);
		//	c.banks = banks1;
			
		//}
	}
	
	boolean flag = false;
	public synchronized void takeloan() throws InterruptedException{
		int famount = this.custamount;
		
		while(this.custamount != 0 && ((this.banks.size()) != 0)) {
			
			Random rand = new Random(); 
	        bank randombank = this.banks.get(rand.nextInt(this.banks.size()));
	        String bankcalled = randombank.getbname();
	       // System.out.println("Calling bank: " + bankcalled);
			
		//	System.out.println("Name is :"+ name + " and amount is:" + amount);
			
			
			
			while (randombank.queue.size() == 0) {
                try {
                	//System.out.println("waiting!\n");
                	this.t2.wait();
                	Thread.sleep(100);
                	}
                catch(Exception e){ 
                	}
                }//while
			
				try {
            	//Thread.sleep(1000);
            	}
				catch(Exception e){ 
				}

				if (randombank.queue.size() == randombank.sizelimit) {
				try {
	            notifyAll(); }
				catch(Exception e){ 
                }
				}
			
				double amtreq = Math.random();
				if(this.custamount >=50) {
				amtreq = amtreq * 50 + 1;}
				else {
					amtreq = amtreq * this.custamount + 1;	
				}
				int amtreqf = (int) amtreq;
				System.out.println(this.custname + " requested for a loan of " + amtreqf + " from bank "+ bankcalled);
				int amountreceived = 0;
				try {
				amountreceived = (int)randombank.queue.remove(0);}
				catch(Exception e){ 
                }
				//System.out.println("For customer "+ name + " Amount received is from bank:"+ bankcalled + "is : " +amountreceived);
				if (amountreceived >= amtreqf) {
				int amountbank = amountreceived - amtreqf;
				System.out.println(bankcalled + " approved a loan of " + amtreqf + " dollar(s) for "+ this.custname);
				this.custamount = this.custamount - amtreqf;
				//System.out.println("approve"+this.custamount+":"+this.custname);
				try {
				randombank.queue.add(amountbank);
				
				}
                catch(Exception e){
                }
				
				}else {
					try {
					System.out.println(bankcalled + " rejected a loan of " + amtreqf + " dollar(s) for "+ this.custname);
					randombank.queue.add(amountreceived);
					//System.out.println(bankcalled + " before removed from list for " + this.custname + "banks: " + this.banks);
					List<bank> bat = this.banks;
					bat.remove(randombank);
					this.banks = bat;
					//System.out.println("reject"+this.custamount+":"+this.custname);
					//System.out.println(bankcalled + " after removed from list for " + this.custname + "banks: " + this.banks);
					//System.out.println("bank list is" + bank);
					if (this.banks.size() == 0) {
						
						int damt = famount - this.custamount;
						
						count = count -1;
						//System.out.println("after reject count is "+ count);
						System.out.println("\n" + this.custname + " was only able to borrow " + damt + " dollar(s). Boo Hoo!\n");
						//System.out.println("\n Amount goal could not reached for " + name + " Boo hoo!");
					}
						}
					catch(Exception e){
					}
				}
				//System.out.println("\nAmount left for " + this.custname + " is: " + this.custamount);	
			
			//Thread.sleep(1000);	
			
		}//while
	
		//System.out.println("Out while loop :" + this.custamount + ":" +this.banks.size() + ":"+ this.custname);
		if (this.custamount == 0) {
		System.out.println("\n" + this.custname + " has reached the objective of " + famount + " dollar(s). Woo Hoo!\n");
		count = count-1;
		//System.out.println("after goal count is "+ count);
		}
		
		if (count == 0) {
			//System.out.println("in count = 0");
			for(int i= 0 ;i < banks1.size(); i++) {
				
				bank r = banks1.get(i);
				String s = r.getbname();
				int amt = 0;
				try {
				amt = (int)r.queue.remove(0);}
				catch(Exception e){
				}
				System.out.println("\n" +s + " has " + amt + " dollar(s) remaining.");
				
			}
		}
		
	}//takeloan
	

	@Override
	public void run() {
		//System.out.println("In customer "+ this.custname+ ":" +this.custamount);
		synchronized(this) {
		try
        { 
            this.takeloan(); 
        } 
        catch(InterruptedException e) 
        { 
            e.printStackTrace(); 
        }
		// TODO Auto-generated method stub
		}
	}
	
}//bankclass


